{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Script to Design Combinatorial Library of Linkers for Grafting EF-hand motif onto Loop 7 of mFAP2b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def revcomp(seq):\n",
    "    return seq.translate(str.maketrans('ACGTacgtRYMKrymkVBHDvbhd', 'TGCAtgcaYRKMyrkmBVDHbvdh'))[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "codons = {\n",
    "            'A': 'GCG',\n",
    "            'C': 'TGC',\n",
    "            'D': 'GAT',\n",
    "            'E': 'GAA',\n",
    "            'F': 'TTT',\n",
    "            'G': 'GGC',\n",
    "            'H': 'CAT',\n",
    "            'I': 'ATT',\n",
    "            'K': 'AAA',\n",
    "            'L': 'CTG',\n",
    "            'M': 'ATG',\n",
    "            'N': 'AAC',\n",
    "            'P': 'CCG',\n",
    "            'Q': 'CAG',\n",
    "            'R': 'CGT',\n",
    "            'S': 'AGC',\n",
    "            'T': 'ACC',\n",
    "            'V': 'GTG',\n",
    "            'W': 'TGG',\n",
    "            'Y': 'TAT',\n",
    "            '*': 'TAA'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def oligo_sequences(start, end, target, out):\n",
    "\n",
    "    target_dna = 'GACAAAGATGGTGACGGCTATATTTCCGCCGCCGAAGCAGCAGCTCAG'\n",
    "\n",
    "    overlap_5prime =  'AGTTCAAATGGGACGCA'\n",
    "    overlap_3prime = 'CTGACCGGTACCCTGC'\n",
    "\n",
    "    oligos_to_order = []\n",
    "    for i in out:\n",
    "        \n",
    "        raw_protein_insert = i.split(start)[-1].split(end)[0]\n",
    "        n_protein_insert = raw_protein_insert.split(target)[0]\n",
    "        c_protein_insert = raw_protein_insert.split(target)[-1]\n",
    "        \n",
    "        n_dna_insert = []\n",
    "        if len(n_protein_insert) > 0:\n",
    "            for codon in n_protein_insert:\n",
    "                n_dna_insert.append(codons[codon])\n",
    "        else:\n",
    "            n_dna_insert.append('')\n",
    "\n",
    "        c_dna_insert = []\n",
    "        if len(c_protein_insert) > 0:\n",
    "            for codon in c_protein_insert:\n",
    "                c_dna_insert.append(codons[codon])\n",
    "        else:\n",
    "            c_dna_insert.append('')\n",
    "\n",
    "        tmp_oligo = overlap_5prime+''.join(n_dna_insert)+target_dna+''.join(c_dna_insert)+overlap_3prime\n",
    "        oligos_to_order.append(tmp_oligo)\n",
    "\n",
    "    return oligos_to_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of oligos before removing redundant sequences = 576\n",
      "Number of oligos after removing redundant sequences = 454\n",
      "Maximum oligo length = 60\n",
      "Maximum oligo length = 60\n",
      "Number of forward oligos with redundancy = 454\n",
      "Number of reverse oligos with redundancy = 454\n",
      "Number of forward oligos without redundancy = 30\n",
      "Number of reverse oligos without redundancy = 38\n",
      "\n",
      "Forward Oligonucleotides:\n",
      "Name\t\t\tSequence\n",
      "QFKWDAQFDKDGDGYISA\tAGTTCAAATGGGACGCACAGTTTGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAEGDKDGDGYISA\tAGTTCAAATGGGACGCAGAAGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAEDKDGDGYISA\tAGTTCAAATGGGACGCAGAAGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDATTDKDGDGYISA\tAGTTCAAATGGGACGCAACCACCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHVGDKDGDGYISA\tAGTTCAAATGGGACGCACATGTGGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAQFGDKDGDGYISA\tAGTTCAAATGGGACGCACAGTTTGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHDKDGDGYISA\tAGTTCAAATGGGACGCACATGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAEYKDKDGDGYISA\tAGTTCAAATGGGACGCAGAATATAAAGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDADKDGDGYISA\tAGTTCAAATGGGACGCAGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDATTKGDKDGDGYISA\tAGTTCAAATGGGACGCAACCACCAAAGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHGDKDGDGYISA\tAGTTCAAATGGGACGCACATGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHVKDKDGDGYISA\tAGTTCAAATGGGACGCACATGTGAAAGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAGDKDGDGYISA\tAGTTCAAATGGGACGCAGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHYDKDGDGYISA\tAGTTCAAATGGGACGCACATTATGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHYKDKDGDGYISA\tAGTTCAAATGGGACGCACATTATAAAGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAEYDKDGDGYISA\tAGTTCAAATGGGACGCAGAATATGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDATDKDGDGYISA\tAGTTCAAATGGGACGCAACCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHYKGDKDGDGYISA\tAGTTCAAATGGGACGCACATTATAAAGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAQGDKDGDGYISA\tAGTTCAAATGGGACGCACAGGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHVDKDGDGYISA\tAGTTCAAATGGGACGCACATGTGGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAQFKDKDGDGYISA\tAGTTCAAATGGGACGCACAGTTTAAAGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAEYKGDKDGDGYISA\tAGTTCAAATGGGACGCAGAATATAAAGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDATGDKDGDGYISA\tAGTTCAAATGGGACGCAACCGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDATTGDKDGDGYISA\tAGTTCAAATGGGACGCAACCACCGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHYGDKDGDGYISA\tAGTTCAAATGGGACGCACATTATGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAQFKGDKDGDGYISA\tAGTTCAAATGGGACGCACAGTTTAAAGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAHVKGDKDGDGYISA\tAGTTCAAATGGGACGCACATGTGAAAGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAEYGDKDGDGYISA\tAGTTCAAATGGGACGCAGAATATGGCGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDATTKDKDGDGYISA\tAGTTCAAATGGGACGCAACCACCAAAGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "QFKWDAQDKDGDGYISA\tAGTTCAAATGGGACGCACAGGACAAAGATGGTGACGGCTATATTTCCGCCG\n",
      "\n",
      "Reverse Oligonucleotides:\n",
      "Name\t\t\tSequence\n",
      "GYISAAEAAAQPKSILTGTL\tGCAGGGTACCGGTCAGAATGCTTTTCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGTHILTGTL\tGCAGGGTACCGGTCAGAATATGGGTGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQKSILTGTL\tGCAGGGTACCGGTCAGAATGCTTTTCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPGRLTGTL\tGCAGGGTACCGGTCAGACGGCCCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPTNILTGTL\tGCAGGGTACCGGTCAGAATGTTGGTCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQNILTGTL\tGCAGGGTACCGGTCAGAATGTTCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGSILTGTL\tGCAGGGTACCGGTCAGAATGCTGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGTNILTGTL\tGCAGGGTACCGGTCAGAATGTTGGTGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGKILTGTL\tGCAGGGTACCGGTCAGAATTTTGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPTHILTGTL\tGCAGGGTACCGGTCAGAATATGGGTCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGKSILTGTL\tGCAGGGTACCGGTCAGAATGCTTTTGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGHGRLTGTL\tGCAGGGTACCGGTCAGACGGCCATGGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQTNILTGTL\tGCAGGGTACCGGTCAGAATGTTGGTCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQHGRLTGTL\tGCAGGGTACCGGTCAGACGGCCATGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQRLTGTL\tGCAGGGTACCGGTCAGACGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQHILTGTL\tGCAGGGTACCGGTCAGAATATGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQKILTGTL\tGCAGGGTACCGGTCAGAATTTTCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPLTGTL\tGCAGGGTACCGGTCAGCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPSILTGTL\tGCAGGGTACCGGTCAGAATGCTCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPILTGTL\tGCAGGGTACCGGTCAGAATCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQILTGTL\tGCAGGGTACCGGTCAGAATCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGLTGTL\tGCAGGGTACCGGTCAGGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQSILTGTL\tGCAGGGTACCGGTCAGAATGCTCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPNILTGTL\tGCAGGGTACCGGTCAGAATGTTCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGTKILTGTL\tGCAGGGTACCGGTCAGAATTTTGGTGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPRLTGTL\tGCAGGGTACCGGTCAGACGCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGHILTGTL\tGCAGGGTACCGGTCAGAATATGGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGRLTGTL\tGCAGGGTACCGGTCAGACGGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPHILTGTL\tGCAGGGTACCGGTCAGAATATGCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPTKILTGTL\tGCAGGGTACCGGTCAGAATTTTGGTCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGILTGTL\tGCAGGGTACCGGTCAGAATGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGGRLTGTL\tGCAGGGTACCGGTCAGACGGCCGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPKILTGTL\tGCAGGGTACCGGTCAGAATTTTCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQGNILTGTL\tGCAGGGTACCGGTCAGAATGTTGCCCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQTKILTGTL\tGCAGGGTACCGGTCAGAATTTTGGTCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQPHGRLTGTL\tGCAGGGTACCGGTCAGACGGCCATGCGGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQLTGTL\tGCAGGGTACCGGTCAGCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "GYISAAEAAAQTHILTGTL\tGCAGGGTACCGGTCAGAATATGGGTCTGAGCTGCTGCTTCGGCGGCGGAAATATAGC\n",
      "\n",
      "Total number of oligo combinations = 1140\n"
     ]
    }
   ],
   "source": [
    "query = ['QFKQNDNPRYHGSTNI',\n",
    "         'HVKSNDSPSLSGNTKI',\n",
    "         'QFKKDDDPRYHGSTHI',\n",
    "         'EYKKNDDPRLKGSKSI',\n",
    "         'HYKTNDYPSLNGSKSI',\n",
    "         'TTKGENDFHGR'\n",
    "        ]\n",
    "\n",
    "target = 'DKDGDGYISAAEAAAQ' # EF-hand motif\n",
    "\n",
    "start = 'MASSHHHHHHSSGLVPRGSSMSRAAQLLPGTWQVTMTNEDGQTSQGQWHFQPRSPYTMDIVAQGTISDGRPIVGYGKATVKTPDTLDIDITYPSLGNIKAQGQITMDSPTQFKWDA'\n",
    "end = 'LTGTLQRQE'\n",
    "\n",
    "G_ = ['', 'G']\n",
    "G_P_ = ['', 'G', 'P']\n",
    "\n",
    "new_sequences = []\n",
    "for q in query:\n",
    "    for N_term in G_:\n",
    "        for Cterm in G_P_:\n",
    "            for num_former in range(0, 4):\n",
    "                for num_latter in range(0, 4):\n",
    "                    if num_latter == 0:\n",
    "                        tmp = start+q[:num_former]+N_term+target+Cterm+q[-num_latter:0]+end\n",
    "                    else:\n",
    "                        tmp = start+q[:num_former]+N_term+target+Cterm+q[-num_latter:]+end\n",
    "                    new_sequences.append(tmp)\n",
    "\n",
    "print('Number of oligos before removing redundant sequences = {}'.format(len(new_sequences)))\n",
    "protein_list = list(set(new_sequences))\n",
    "print('Number of oligos after removing redundant sequences = {}'.format(len(protein_list)))\n",
    "\n",
    "\n",
    "oligo_list = oligo_sequences(start, end, target, protein_list)\n",
    "\n",
    "overlapping_region = 'GCTATATTTCCGCCG'\n",
    "\n",
    "dic = {}\n",
    "\n",
    "test = zip(protein_list, oligo_list)\n",
    "for i in test:\n",
    "    coding_dna = Seq('C'+i[1]+'AG', generic_dna)\n",
    "    protein_string = str(coding_dna.translate())\n",
    "    assert(protein_string in i[0])\n",
    "\n",
    "\n",
    "    oligo_name = i[0].split(start)[-1].split(end)[0]\n",
    "    # Now split into two oligos for ameliorated pricing\n",
    "    oligo_9_fwd = i[1].split(overlapping_region)[0]+overlapping_region\n",
    "    oligo_10_fwd = overlapping_region+i[1].split(overlapping_region)[-1]\n",
    "    oligo_10_rev = revcomp(oligo_10_fwd)\n",
    "\n",
    "    dic[oligo_name] = (oligo_9_fwd, oligo_10_rev)\n",
    "\n",
    "\n",
    "print('Maximum oligo length = {}'.format(max([len(v[0]) for k,v in dic.items()])))\n",
    "print('Maximum oligo length = {}'.format(max([len(v[1]) for k,v in dic.items()])))\n",
    "\n",
    "fwd = []\n",
    "rev = []\n",
    "\n",
    "for k,v in dic.items():\n",
    "    fwd.append(v[0])\n",
    "    rev.append(v[1])\n",
    "\n",
    "print('Number of forward oligos with redundancy = {}'.format(len(fwd)))\n",
    "print('Number of reverse oligos with redundancy = {}'.format(len(rev)))\n",
    "\n",
    "fwd_unique = list(set(fwd))\n",
    "rev_unique = list(set(rev))\n",
    "\n",
    "print('Number of forward oligos without redundancy = {}'.format(len(fwd_unique)))\n",
    "print('Number of reverse oligos without redundancy = {}'.format(len(rev_unique)))\n",
    "\n",
    "print('\\nForward Oligonucleotides:')\n",
    "print('Name\\t\\t\\tSequence')\n",
    "for f in fwd_unique:\n",
    "    name_ = 'C'+f[:-1]\n",
    "    name = str(Seq(name_, generic_dna).translate())\n",
    "    print(name+'\\t'+f)\n",
    "\n",
    "print('\\nReverse Oligonucleotides:')\n",
    "print('Name\\t\\t\\tSequence')\n",
    "for r in rev_unique:\n",
    "    name_ = 'G'+revcomp(r)[:-1]\n",
    "    name = str(Seq(name_, generic_dna).translate())\n",
    "    print(name+'\\t'+r)\n",
    "\n",
    "print('\\nTotal number of oligo combinations = {}'.format(len(fwd_unique)*len(rev_unique)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:PyRosettaScripts]",
   "language": "python",
   "name": "conda-env-PyRosettaScripts-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
